### ISSUE TITLE
URL-Grabber: parse response code 429 Too Many Requests

### ISSUE THREAD

**Issue by ggzengel** [sabnzbd/sabnzbd#707]
Sabnzbd didn't wait 51 minutes for next retry even if there is a retry-after value. It waits only 60 seconds.

I get:
```

			<html>
			<head>
				<title>Too Many Requests</title>
			</head>

			<body>
				<h1>Too Many Requests</h1>

				<p>Wait 51 minutes or risk being temporarily banned.</p> 

			</body>
			</html>
```

In Logs:
```
2016-10-11 03:38:18,719::INFO::[urlgrabber:101] Grabbing URL https://<snip>
2016-10-11 03:38:19,059::DEBUG::[urlgrabber:119] Error "http error 429: too many requests" trying to get the url https://<snip>
2016-10-11 03:38:19,113::DEBUG::[urlgrabber:276] No usable response from indexer, retry after 60 sec
2016-10-11 03:38:19,113::INFO::[urlgrabber:188] Retry URL https://<snip>
```

Response:
```
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Connection: close
Content-Length: 196
Content-Type: text/html; charset=utf-8
Date: Tue, 11 Oct 2016 03:32:04 GMT
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Pragma: no-cache
Retry-After: 3035
Server: Apache/2.2.27 (Unix) mod_ssl/2.2.27 OpenSSL/1.0.1e-fips mod_bwlimited/1.4 PHP/5.4.31
X-Powered-By: PHP/5.4.31
```

**Comment by Safihre** [comment]
Will look into what's going wrong. Could you maybe send me an example URL? Safihre@sabnzbd.org 
I will only try it one or twice to check the headers!

**Comment by thezoggy** [comment]
the code was to handle newznab sites.. not all sites..

**Comment by thezoggy** [comment]
for nzbfinder and other sites that send the header `retry-after` we use that. 
https://github.com/sabnzbd/sabnzbd/commit/262b1304f45c22c137f7acb69a0009389609fe85

now for other sites, we only process the data if a 200 code is returned.. otherwise we try again after 60 secs... 

now for your response, `Retry-After: 3035` 
3035 *should* be what is parsed and used...  assuming it reaches that piece of code

**Comment by thezoggy** [comment]
honestly the whole urlgrabber code could be refactored into like 3 lines of Request code...  really shouldnt have to much with urllib/urllib2/etc

**Comment by thezoggy** [comment]
https://github.com/sabnzbd/sabnzbd/blob/develop/sabnzbd/urlgrabber.py#L275
I would suspect dropping `200` from that check would solve everything.. since the site is passing 429..   prob a better choice would be to explicit about what codes you dont want to proceed with..

**Comment by Safihre** [comment]
@thezoggy, I just simulated it and @ggzengel is right. The problem is not the code you mention, but that when ```urllib2.urlopen``` encounters a 429 it will throw an ```urllib2.HTTPError``` and the code will not continue.
The solution would be, like in @ggzengel's suggestion, to also parse the headers of an failed request:
```python
                try:
                    fn = urllib2.urlopen(req)
                except urllib2.HTTPError, e:
                    fn = e
                except:
```
Would that result in problems?

**Comment by Safihre** [comment]
Actually, here is where we should parse the HTTP error codes and check for the retry header.

**Comment by Safihre** [comment]
Found the "right" way to fix it, will probably be fixed in SABnzbd 1.1.1.
See: eb94212

**Comment by ggzengel** [comment]
This patch works for me. Nice solution.
Thank you.
